core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -109,6 +109,7 @@ public ThisT set(String property, String value) {
   }
 
   protected PartitionSpec writeSpec() {
+    Preconditions.checkState(spec != null, "No data or delete files have been added.");
```

Comment:
We normally note the operation that is failing, like "Cannot determine partition spec: no data or delete files have been added". Also, no need for ending punctuation.